staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -18,19 +18,83 @@ package workqueue
 
 import (
 	"container/heap"
+	"fmt"
+	"strconv"
+	"strings"
 	"sync"
 	"time"
 
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/utils/clock"
 )
 
+// queueLogger is any type that provides the Log and Logf methods.
+type queueLogger interface {
+	Log(args ...any)
+	Logf(format string, args ...any)
+}
+
+// DefaultLogger is set to nil which disables logging but can be replaced
+// by valid queueLogger to enable the logging of workqueue events.
+// For example set it to your *testing.T to get queue events logged in your tests.
+var DefaultLogger queueLogger = nil
```

Comment:
Hmm this is a tough one, not sure how easy it could be to set it accidentally, given that you need to know the variable and the package.  Whilst I take your point that it would be ideal if you can call NewDelayingQueueWithCustomClockAndLogger() it is not always possible to do that.  I created the package enabled way of switching it on precisely because I needed to investigate a failure in someone else's test - where their instantiation of the queue was buried deep in their own controller implementation where it was not possible to swap the call to NewDelayingQueueWithCustomClockAndLogger().  Using this method I was able to turn on the queue logging within just the failing test with one statement and without having to alter their code.

With regards to security, if this is an issue we can remove it.  We could remove the logging entirely but during the development of this PR I have added and removed it so many times that I've concluded that there is value in keeping it.

Reply:
Personally I would keep this because of the utility it provides to the development and testing of the queue code unless there is a very strong reason why it needs to be removed or changed.

Reply:
I've removed **NewDelayingQueueWithCustomClockAndLogger()** and added a **NewTestDelayingQueueWithCustomClockAndLogger()** which is only callable from test suits.  This should prevent people from using the logger outside of testing.

Reply:
> With regards to security, if this is an issue we can remove it. We could remove the logging entirely but during the development of this PR I have added and removed it so many times that I've concluded that there is value in keeping it.

hm, I understand it has debugging value, but I think it would be preferable to either fully expose is it as an API or not at all. The global variable would become part of the API as well and it might not be so easy to remove.

If we only need it for debugging purposes in the workqueue package, we could name it debugLogger and make the variable (DefaultLogger) package private or just use the private constructor from the tests.

If we expect this to be consumed by other people, we could start a discussion on how to do that. We could also consider other approaches like metrics. 

Nevertheless, the observability could be done as a followup.

Reply:
Have removed it.